src/test/java/org/junitpioneer/jupiter/cartesian/CartesianTestExtensionTests.java

```diff
@@ -887,6 +926,25 @@ void shouldHaveTestInfo(@CartesianTest.Values(ints = { 1, 2 }) int i,
 
 	}
 
+	static class CustomCartesianArgumentsProviderTestCases {
+
+		@CartesianTest
+		@ReportEntry("{0}({1})")
+		void twoCustomCartesianArgumentProviders(
+				@CartesianArgumentsSource(FirstCustomCartesianArgumentsProvider.class) String string,
+				@CartesianArgumentsSource(SecondCustomCartesianArgumentsProvider.class) int integer) {
```

Comment:
Looks good, although maybe it's better if we take the same route as with the other annotations and make this an inner class of `@CartesianTest`, e.g. `@CartesianTest.Source`, then users can use either of these two variants:

```java
@CartesianTest.Source(FirstCustomCartesianArgumentsProvider.class) String string,
@Source(SecondCustomCartesianArgumentsProvider.class) int integer
```

Reply:
What about `@CartesianTest.Arguments`? Similar pattern to `@CartesianEnumSource` --> `@CartesianTest.Enum` and `@CartesianValueSource` --> `@CartesianTest.Values`

Reply:
I understand your reasoning @nipafx and @scordio. I watched the stream and read the other issue where you decided about the inner classes.

However, I personally am more in favour of the current name `@CartesianArgumentsSource`. The reason for that is that people are already familiar with `ArgumentsSource`. In addition to that the target audience for the `@CartesianArgumentsSource` annotation is other library developers that would like to provide their own `CartesianArgumentsProvider`. The example in the test case is just an example of a potential use case. 

From an API point of view I think that it is better to keep it close to what people are used to. If it really has to be an inner class I think using `@CartesianTest.ArgumentsSource`,  i.e indicating an `ArgumentsSource` for `CartesianTest` fits better, although it is longer then `@CartesianArgumentsSource` and it will clash in a standalone mode.


`@CartesianTest.Source` or `@CartesianTest.Arguments` feels more like an end user annotations where people will provide the `source` or `arguments` values themselves, i.e. like in `@CartesianTest.Values`.

Reply:
Good argument, I'm convinced.